lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+
+        # Temporary storage for SendChannels used in the `stream` API.  Uses a
+        # `WeakSet` to automate cleanup.
+        self._stream_channels = collections.defaultdict(weakref.WeakSet)
+        self._pending_requests = {}
+        self._sync_handlers = collections.defaultdict(list)
+        self._async_handlers = collections.defaultdict(list)
+
+        self._run_lock = trio.Lock()
+
+        # Signal when a new remote connection is established
+        self._remote_connections_changed = trio.Condition()  # type: ignore
+
+        # Signal when at least one remote has had a subscription change.
+        self._remote_subscriptions_changed = trio.Condition()  # type: ignore
+
+        # events used to signal that the endpoint has fully booted up.
+        self._message_processing_loop_running = trio.Event()
+        self._connection_loop_running = trio.Event()
+        self._process_broadcasts_running = trio.Event()
+
+        # internal signal that local subscriptions have changed.
+        self._subscriptions_changed = trio.Event()
+
+        self._server_running = trio.Event()
+        self._server_stopped = trio.Event()
+
+    #
+    # Running and endpoint
+    #
+    @property
+    def is_running(self) -> bool:
+        return not self.is_stopped and self._running.is_set()
+
+    @property
+    def is_stopped(self) -> bool:
+        return self._stopped.is_set()
+
+    async def wait_started(self) -> None:
+        await self._running.wait()
+
+    async def wait_stopped(self) -> None:
+        await self._stopped.wait()
+
+    @asynccontextmanager
+    async def run(self) -> AsyncGenerator[EndpointAPI, None]:
+        async with trio.open_nursery() as nursery:
+            await self._start(nursery)
+            try:
+                yield self
+            finally:
+                await self._stop()
+
+    async def _start(self, nursery: trio_typing.Nursery) -> None:
+        if self.is_running:
+            raise LifecycleError("Endpoint is already running")
+        elif self.is_stopped:
+            raise LifecycleError("Endpoint has already been run and stopped")
+
+        nursery.start_soon(self._run)
+        # this is a trade-off between allowing flexibility in how long it takes
+        # for a `Runnable` to start and getting good errors in the event that
+        # the `_run` method forgets to set the `_running` event.
+        with trio.fail_after(2):
+            await self.wait_started()
+
+    async def _stop(self) -> None:
+        if self.is_stopped:
+            return
+        self._stopped.set()
+        await self._cleanup()
+
+    async def _run(self) -> None:
+        # This send/receive channel is fed by connected `RemoteEndpoint`
+        # objects which feed received events into the send side of the channel.
+        # Those messages are then retrieved from the receive channel by the
+        # `_process_received_messages` daemon.
+        (self._message_send_channel, message_receive_channel) = cast(
```

Comment:
The fact that trio gives us two channels makes naming hard, but `_message_send_channel` does not sound like the channel we receive messages with, it took me a bit to figure out what it was for. Maybe `_incoming_message_channel_send` and `_incoming_message_channel_receive`?

Reply:
I'll see if I can find better names.  Agreed on this being a confusing one.

Reply:
I renamed this to `_inbound_send_channel` (and there is a new `_outbound_send_channel` to signal the direction of message flow.